ASM long -> 3 int transformation for light engine

BelgianSalamander opened the pull request.

Niko-sk2x committed:
Add processCubeUnloads to ASM transformer

Niko-sk2x committed:
Add scheduleCubeUnload to ASM transformer

Niko-sk2x committed:
Add saveAllCubes to ASM transformer, prepare for adding cubeSave

Niko-sk2x committed:
Add cubeSave to ASM transformer

CursedFlames committed:
ASM to change computedLevels and queues in DynamicGraphMinFixedPoint to type Object

BelgianSalamander committed:
Added basic remapping

BelgianSalamander committed:
Fixed method remapping

BelgianSalamander committed:
Added data flow analysis

BelgianSalamander committed:
SkyLightEngine now remaps fine

BelgianSalamander committed:
Created LinkedInt3HashSet. Still very buggy

DaMatrix committed:
implement a fast hashmap for 3D integer points to unsigned byte values

DaMatrix committed:
add poll method to Int3UByteLinkedHashMap

BelgianSalamander committed:
Fixed LinkedInt3HashSet

BelgianSalamander committed:
Added LinkedInt3HashSet to DynamicGraphMinFixedPoint

BelgianSalamander committed:
Added removeIf, MinPoint abstract methods and annotations

BelgianSalamander committed:
Nearly finished DynamicGraphMinFixedPoint transform

BelgianSalamander committed:
Added TypeTransformer system

BelgianSalamander committed:
Started fixing DUP and SWAP

BelgianSalamander committed:
Fixed new getSyntheticEmitter

BelgianSalamander committed:
Created TypeTransformerMethods test

BelgianSalamander committed:
Can now load pre-transformed classes from .mixin.out

BelgianSalamander committed:
Added basic support for mixin's invokers

BelgianSalamander committed:
Fixed unit test

BelgianSalamander committed:
Added interface system

BelgianSalamander committed:
Added variables to JSONBytecodeFactory. Worlds load!!!

BelgianSalamander committed:
Fixed last warning!

BelgianSalamander committed:
Removed class duplication

BelgianSalamander committed:
Add ConstructorReplacer

BelgianSalamander committed:
Replaced interface system with better invoker replacement

BelgianSalamander committed:
Small cleanup

BelgianSalamander committed:
Temporary fix for CubicAquifer ArrayIndexOutOfBoundsException for large coordinates.

BelgianSalamander committed:
Fixed checkstyle syntax issues

BelgianSalamander committed:
Passes CheckStyle

BelgianSalamander committed:
Documentation

BelgianSalamander committed:
Fixed something from previous commit

BelgianSalamander committed:
Added type inference unit test

BelgianSalamander committed:
Added test for config

BelgianSalamander committed:
Merge branch 'MC_1.17' into MC_1.17_lightEngineAsm

BelgianSalamander committed:
Fixed mistake from merge

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/levelgen/aquifer/CubicAquifer.java:
```diff
@@ -91,7 +91,16 @@ private int getIndex(int x, int y, int z) {
+        int index = (localY * this.gridSizeZ + localZ) * this.gridSizeX + localX;
+
+        //FIXME: Because of the way the cache is implemented, this can cause an ArrayIndexOutOfBoundsException.
+        //The coordinates are unpacked from longs so when outside the proper range, they wrap around and this returns nonsense values.
```
Does this happen in cubic or vanilla worlds? If it only happens in vanilla, this check should ensure that we are in a cubic world and raise an exception instead.

BelgianSalamander replied:
The reason BlockPos wraps around within vanilla bounds is because of [../mixin/core/common/MixinBlockPos](https://github.com/OpenCubicChunks/CubicChunks2/blob/a83bb3acf7221ecbaf9b7d492d0fde94967a4158/src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/MixinBlockPos.java) which results in less bits being used in a long to represent those coordinates whatever the world type. Therefore, this wraparound will most likely occur within any world using CubicAquifer.

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/progress/MixinChunkProgressListener.java:
```diff
@@ -0,0 +1,9 @@
+import org.spongepowered.asm.mixin.Mixin;
+
+@Mixin(ChunkProgressListener.class)
+public interface MixinChunkProgressListener extends CubeProgressListener {
```
What is this interface supposed to achieve?

Cyclonit reviewed (commented):
If IChunkMapInternal.java is supposed to be an empty file, can it be deleted?

While I'd like to claim that I understand more of this PR, most of it is black magic to me. I fear its up to Barteks to review this one properly.

BelgianSalamander committed:
Start adding info to crash log

BelgianSalamander commented:
I will probably be re-writing/simplifying a big portion of this soon.

BelgianSalamander committed:
Merge 1.18 into ASM

BelgianSalamander committed:
Automatically map the tt config

BelgianSalamander committed:
Get type info at compile time

BelgianSalamander committed:
Actually load type info

BelgianSalamander committed:
Worlds load!

BelgianSalamander committed:
Fix Aquifer

BelgianSalamander committed:
Much needed clean-up

BelgianSalamander committed:
Make tests work

BelgianSalamander committed:
Merge with upstream

BelgianSalamander committed:
Fix loom issues

CursedFlames commented on .idea/codeStyles/Project.xml:
```diff
@@ -107,7 +107,7 @@
       <option name="WRAP_LONG_LINES" value="true" />
       <indentOptions>
-        <option name="INDENT_SIZE" value="4" />
+        <option name="TAB_SIZE" value="2" />
```
unnecessary?

CursedFlames commented on buildSrc/src/main/java/io/github/opencubicchunks/gradle/TypeTransformConfigGen.java:
```diff
@@ -0,0 +1,397 @@
+        }
+    }
+
+    /*public static String apply(Project project, String content) throws IOException {
```
remove commented code

CursedFlames commented on config/checkstyle/checkstyle.xml:
```diff
@@ -160,7 +160,11 @@
+        <module name="CyclomaticComplexity">
+            <property name="switchBlockAsSingleDecisionPoint" value="true"/>
+            <property name="max" value="15"/>
+            <property name="tokens" value="LITERAL_WHILE , LITERAL_DO , LITERAL_FOR , LITERAL_IF , LITERAL_SWITCH , LITERAL_CASE , LITERAL_CATCH , QUESTION"/>
```
not sure of the consequences of this; would like others to review

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/utils/Int3HashSet.java:
empty file?

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/utils/LinkedInt3HashSet.java:
```diff
@@ -0,0 +1,640 @@
+                            first = bucketAddr;
+                        }
+
+                        //If last is set, set the the last value's pointer to point here and set this pointer to last
```
"the the" -> "the"

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/utils/LinkedInt3HashSet.java:
```diff
@@ -0,0 +1,640 @@
+        this.last = 0L;
+    }
+
+    //Cached index of value
```
this is a very vague comment, please edit to clarify

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/utils/LinkedInt3HashSet.java:
```diff
@@ -0,0 +1,640 @@
+        getFirstSetBitInFirstBucket(0);
+    }
+
+    protected void getFirstSetBitInFirstBucket(int start) {
```
argument is unused?

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/utils/Int3List.java:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    public void set(int index, int x, int y, int z) {
+        if (index >= this.size) {
```
index < 0?
(same for other methods in this file)

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/utils/Int3List.java:
```diff
@@ -0,0 +1,308 @@
+
+    public boolean remove(int x, int y, int z) {
+        for (int index = 0; index < size; index++) {
+            if (getX(index) == x && getY(index) == y && getZ(index) == z) {
```
could remove duplicate bounds checks by inlining these but it probably doesn't matter

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/utils/Int3UByteLinkedHashMap.java:
```diff
@@ -0,0 +1,808 @@
+ *
+ * @author DaPorkchop_
+ */
+public class Int3UByteLinkedHashMap implements AutoCloseable {
```
reviewed Salamander's changes; assuming that Pork's original version is fine

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/IChunkMapInternal.java:
empty file?

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/bytecodegen/JSONBytecodeFactory.java:
```diff
@@ -0,0 +1,339 @@
+
+        int opcode = opcodeFromName(insnName);
+
+        return (insnList, indexes) -> insnList.add(new InsnNode(opcode));
```
rename `indexes` to `__` for consistency

(in general, looks like the indices are only used for var instructions, assuming I'm reading the code correctly?)

CursedFlames commented on src/test/java/io/github/opencubicchunks/cubicchunks/utils/Int3UByteLinkedHashMapTest.java:
```diff
@@ -0,0 +1,263 @@
+
+    @Test
+    public void testIterators() {
+        /*try(Int3UByteLinkedHashMap map = new Int3UByteLinkedHashMap()){
```
remove commented code (or finish implementing this test case?)

BelgianSalamander committed:
Remove method alias system - don't know why I added it in the first place and never used it

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformSubtype.java:
```diff
@@ -0,0 +1,464 @@
+        }
+    }
+
+    public enum SubType {
```
rename to "Kind" or similar, as discussed

Niko-sk2x committed:
Bump dasm and javaheaders version to get around jitpack issues

Niko-sk2x committed:
Remap names in ASMConfigPlugin for dasm

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformSubtype.java:
```diff
@@ -0,0 +1,464 @@
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.VarInsnNode;
+
+public class TransformSubtype {
```
`TransformDependentType` or `TransformDerivedType` as possible name alternatives

Alternatively, could rename `TransformType` to `TransformBaseType`, and this to `TransformType`; I think keeping `TransformType` as it is is probably better though.

Niko-sk2x committed:
Fix checkstyle and CubicChunksCore submodule

BelgianSalamander committed:
Start docs

BelgianSalamander committed:
Merge remote-tracking branch 'upstream/MC_1.18' into MC_1.17_lightEngineAsm

# Conflicts:
#	CubicChunksCore
#	src/main/java/io/github/opencubicchunks/cubicchunks/mixin/ASMConfigPlugin.java

BelgianSalamander committed:
Small merge fixes

CursedFlames commented on src/test/java/io/github/opencubicchunks/cubicchunks/typetransformer/TypeTransformerMethods.java:
```diff
@@ -0,0 +1,291 @@
+            "net.minecraft.class_3560", //LayerLightSectionStorage
+            "net.minecraft.class_3547", //BlockLightSectionStorage
+            "net.minecraft.class_3569", //SkyLightSectionStorage
+            "net.minecraft.class_4076",
```
add comments for what the rest of these are?

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/TypeTransformer.java:
```diff
@@ -0,0 +1,2006 @@
+        frames[0] = new Frame<>(numLocals, 0);
+
+        int varIndex = 0;
+        if (!ASMUtil.isStatic(methodNode)) {
```
we assume earlier on that passed methods are abstract (and therefore non-static), no?

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/TypeTransformer.java:
```diff
@@ -0,0 +1,2006 @@
+        }
+
+        int i = 0;
+        for (Type argType : args) {
```
using a `for (element : list)` loop and then also having the loop index variable `i` seems unnecessarily difficult to read - any reason to not simply do `for (int i = 0; i < args.length; i++)` instead?

on a related note, `args` and `argTypes` is kind of unclear; might be worth renaming (this applies to multiple places in this file)

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/TypeTransformer.java:
```diff
@@ -0,0 +1,2006 @@
+     * Transform a method call whose replacement is given in the config
+     *
+     * @param context Transform context
+     * @param methodCall The actual method cal insn
```
"cal" -> "call"

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/TypeTransformer.java:
```diff
@@ -0,0 +1,2006 @@
+                offsets[i][j] += totalSize;
+            }
+
+            totalSize += transform.getTransformType() == null ? args[0].getSize() : transform.getTransformedSize();
```
admittedly I can't fully understand this code by any means, but this `args[0].getSize()` seems strange - should it be `args[i].getSize()`? (if not, might want a comment explaining it)

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/TypeTransformer.java:
```diff
@@ -0,0 +1,2006 @@
+     * @param returnValue The return value of the method call, if the method returns void this should be null
+     * @param args The arguments of the method call. This should include the instance ('this') if it is a non-static method
+     */
+    private void applyDefaultReplacement(TransformContext context, int i, MethodInsnNode methodCall, TransformTrackingValue returnValue, TransformTrackingValue[] args) {
```
`returnValue` should be `@Nullable` then?

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/TypeTransformer.java:
```diff
@@ -0,0 +1,2006 @@
+
+        TypeInfo hierarchy = config.getTypeInfo();
+
+        Type potentionalOwner = types.get(0);
```
"potentional" -> "potential" (make same change elsewhere too)

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/TypeTransformer.java:
```diff
@@ -0,0 +1,2006 @@
+        context.target.instructions.insert(methodCall, replacement);
+        context.target.instructions.remove(methodCall);
+
+        return;
```
unnecessary `return`

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/TypeTransformer.java:
```diff
@@ -0,0 +1,2006 @@
+
+        if ((newMethod.access & Opcodes.ACC_ABSTRACT) != 0) {
+            transformAbstractMethod(newMethod, start, methodID, newMethod, context);
+            return;
```
probably use an if/else with the final println afterwards instead of returning here, to avoid duplicating the println statement

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/TypeTransformer.java:
```diff
@@ -0,0 +1,2006 @@
+     * @param subType The type of synthetic method this is
+     * @param original The original method this is a synthetic version of
+     */
+    private static void markSynthetic(MethodNode methodNode, String subType, String original, String ownerName) {
```
might be clearer to rename `subType` to `kind` or similar

BelgianSalamander committed:
Make type hint specification more sensible

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformSubtype.java:
```diff
@@ -0,0 +1,464 @@
+
+    /**
+     * @return The single transformed type of this transform subtype.
+     * @throws IllegalStateException If this subtype is not NONE or there is not a single transformed type. To get all the types
```
this javadoc seems to be incorrect; the actual logic is error if subtype is NONE and there's not exactly one transformed type.

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformSubtype.java:
```diff
@@ -0,0 +1,464 @@
+        }
+    }
+
+    //Does not work with array dimensionality
```
is this comment still accurate? there seems to be code for handling array dimensionality in `resultingTypes`

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformSubtype.java:
```diff
@@ -0,0 +1,464 @@
+import org.objectweb.asm.tree.VarInsnNode;
+
+public class TransformSubtype {
+    //A reference to a TransformType. THis means when the transform type gets changed all referenced ones can get notified
```
"THis" -> "This"

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformTrackingValue.java:
```diff
@@ -0,0 +1,194 @@
+    private final Set<TransformTrackingValue> valuesWithSameType = new HashSet<>();
+    private final Config config;
+
+    public TransformTrackingValue(@Nullable Type type, AncestorHashMap<FieldID, TransformTrackingValue> fieldPseudoValues, Config config) {
```
Any reason to not just use `this(type, TransformSubtype.createDefault(type), fieldPseudoValues, config)` in place of this constructor's implementation? would reduce code duplication

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformTrackingValue.java:
```diff
@@ -0,0 +1,194 @@
+        this.transform.getTransformTypePtr().setValue(transformType);
+    }
+
+    public void updateType(@Nullable TransformType oldType, TransformType newType) {
```
`oldType` is unused

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformTrackingValue.java:
```diff
@@ -0,0 +1,194 @@
+        return Objects.hash(type, transform);
+    }
+
+    public static <T> Set<T> union(Set<T> first, Set<T> second) {
```
This appears to never be used?

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/config/MethodTransformChecker.java:
```diff
@@ -0,0 +1,90 @@
+        return current.equals(target);
+    }
+
+    public static record Minimum(TransformSubtype returnType, TransformSubtype... parameterTypes) {
```
Might be clearer to rename to `MinimumConditions` or similar

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/AnalysisResults.java:
```diff
@@ -0,0 +1,77 @@
+        TransformSubtype[] types = argTypes;
+        if (!ASMUtil.isStatic(methodNode)) {
+            //If the method is not static then the first element of this.types is the 'this' argument.
+            //This argument is not shown is method descriptors, so we must exclude it
```
"shown is" -> "shown in"

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/AnalysisResults.java:
```diff
@@ -0,0 +1,77 @@
+        TransformSubtype[] argTypes = new TransformSubtype[args.length + offset];
+
+        int idx = 0;
+        for (int i = 0; idx < argTypes.length; i += frames[0].getLocal(i).getSize()) {
```
Very minor, but I think this would be slightly more readable with `for (int idx = 0; idx < argType.length; idx++)` - with `i += frames[0].getLocal(i).getSize()` inside the loop body instead

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/config/ConfigLoader.java:
```diff
@@ -0,0 +1,671 @@
+            if (indices1.isJsonArray()) {
+                for (int j = 0; j < indices1.getAsJsonArray().size(); j++) {
+                    List<Integer> l = new ArrayList<>();
+                    indices[i][j] = l;
```
You use `[i][j]` here but `[j][i]` below and in `generateDefaultIndices`; I assume `[j][i]` is correct? (if the current code is actually correct, might be worth adding a comment explaining the different indices)

BelgianSalamander replied:
Defintely should be `[i][j]` - the code with `[j][i]` seems to be completely unused and it's probably best if it's removed completely.

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/config/ConfigLoader.java:
```diff
@@ -0,0 +1,671 @@
+            }
+
+            if (defaultCallType != null) {
+                callType = defaultCallType;
```
not sure if this coming after `callType` is set is intended; if it is, naming the argument `defaultCallType` seems a bit misleading when it overrides any existing call type

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/config/Config.java:
```diff
@@ -0,0 +1,99 @@
+    private final Map<String, TransformType> types;
+    private final AncestorHashMap<MethodID, List<MethodParameterInfo>> methodParameterInfo;
+    private final Map<Type, ClassTransformInfo> classes;
+    private final Map<Type, InvokerInfo> invokers;
```
`invokers` is currently unused (and has no getter); is this intentional?

CursedFlames commented on src/main/resources/type-transform.json:
```diff
@@ -0,0 +1,864 @@
+      ]
+    },
+    {
+      "method": "i it/unimi/dsi/fastutil/longs/LongLost#forEach (Ljava/util/function/LongConsumer;)V",
```
`LongList`? (again directly below)

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformTrackingInterpreter.java:
```diff
@@ -0,0 +1,544 @@
+    }
+
+    @Override
+    //Because of the custom Frame (defined in Config$DuplicatorFrame) this method may be called multiple times for the same instruction-value pair
```
`DuplicatorFrame` doesn't seem to exist anywhere - presumably this comment is out of date?

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformTrackingInterpreter.java:
```diff
@@ -0,0 +1,544 @@
+    private TransformTrackingValue methodCallOperation(AbstractInsnNode insn, List<? extends TransformTrackingValue> values, int opcode) {
+        //Create bindings to the method parameters
+        MethodInsnNode methodCall = (MethodInsnNode) insn;
+        Type subType = Type.getReturnType(methodCall.desc);
```
maybe rename to `returnType` or similar? (same in `invokeDynamicOperation`)

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformTrackingInterpreter.java:
```diff
@@ -0,0 +1,544 @@
+    public TransformTrackingValue merge(TransformTrackingValue value1, TransformTrackingValue value2) {
+        if (!Objects.equals(value1.getType(), value2.getType())) {
+            //System.err.println("WARNING: Merge types are not equal");
+            return value1;
```
At a glance this seems unsafe - assuming it's fine, a comment explaining why would be useful

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/analysis/TransformTrackingInterpreter.java:
```diff
@@ -0,0 +1,544 @@
+        Type[] argumentTypes = Type.getArgumentTypes(methodResults.methodNode().desc);
+        Type[] allTypes;
+        if (!ASMUtil.isStatic(methodResults.methodNode())) {
+            //Remove the first element because it represents 'this'
```
Comment doesn't make sense here - maybe something like `add the first element, representing 'this'`

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/typetransformer/transformer/TypeTransformer.java:
```diff
@@ -0,0 +1,2006 @@
+     * @param analysisResults The analysis results for this method that were generated by the analysis phase.
+     * @param instructions The instructions of {@code target} before any transformations.
+     * @param varLookup Stores the new index of a variable. varLookup[insnIndex][oldVarIndex] gives the new var index.
+     * @param variableAllocator The variable manager allows for the creation of new variables.
```
Comment should probably be updated to say "variable allocator" for consistency.

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/util/AncestorHashMap.java:
```diff
@@ -0,0 +1,118 @@
+    @Override
+    public boolean containsKey(Object key) {
+        if (key instanceof Ancestralizable method) {
+            for (Type subType : hierarchy.ancestry(method.getAssociatedType())) {
```
Aren't these super-types, not sub-types? `superType` or `ancestorType` might be clearer than `subType`.

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/transform/util/Ancestralizable.java:
```diff
@@ -0,0 +1,8 @@
+
+import org.objectweb.asm.Type;
+
+public interface Ancestralizable<T extends Ancestralizable<T>> {
```
I feel like this could use a better name, but I don't have suggestions.

(Is "ancestralizable" even a word?)

CursedFlames reviewed (commented):
Don't see any major issues.

* `type-transformer-config-docs.md` still needs to be finished
* the transformer stuff could do with more comments/javadocs, of course; don't necessarily need this before merge though, it can be done later
* there's a bunch of things missing `@Nullable` since you use `NonnullByDefault` - I'm not gonna go through and comment on them all; it's pretty minor.
* It might be worth chucking `List<Integer>[][]` in a record class somewhere with a javadoc explaining what the indices actually *do* - I've read this entire PR and I'm still not sure what they are.
* "indices" and "indexes" seem to be used inconsistently. Very minor
* can the int3 list/map/set/etc in `cubicchunks/utils` be moved to cubicchunkscore?

To do before merge:
- [x] Salamander addresses my comments
- [ ] Salamander finishes documentation in `type-transformer-config-docs.md`
- [ ] Would like others to carefully review the following:
    * The int3 list/map/set stuff in `cubicchunks/utils`
    * `TransformTrackingInterpeter.invokeDynamicOperation`
    * `TypeTransformer.transformInvokeDynamicInsn`
   (I'm not confident enough in my knowledge of lambdas and `invokeDynamic` in bytecode; I could have missed issues with the relevant methods.)
- [ ] If possible, would like at least one other person to review this entire PR; I understand that reviewing 11k LoC diff is a big ask though 😅

BelgianSalamander committed:
Start addressing minor review comments

BelgianSalamander committed:
Fix unit test issue

CursedFlames commented:
Also another thing Salamander and I discussed that I forgot to mention here:

For clarity, I think we should use `indices` or similar to refer to regular variable indices, and `offsets` for the case where `long`s and `double`s take two slots rather than one.

BelgianSalamander committed:
More small fixes + renaming

CursedFlames committed:
pass checkstyle

rename static getKind to getKindFor in DerivedTransformType, as otherwise checkstyle complains about overloads that aren't together

CursedFlames committed:
remove unused invokers field on Config

CursedFlames committed:
misc minor tweaks

CursedFlames committed:
Int3List bounds checking + misc
